File: ProblemStatements/0297. SerializeAndDeserializeBinaryTree/serializeAndDeserialize.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Codec {

    // Encodes a tree to a single string.
    public String serialize(TreeNode root) {
        StringBuilder sb = new StringBuilder();
        helper1(root, sb);
        return sb.toString();
    }

    private void helper1(TreeNode node, StringBuilder sb){
        if(node == null){
            sb.append("#,");
            return;
        }
        sb.append(node.val).append(",");
        helper1(node.left, sb);
        helper1(node.right, sb);
    }

    // Decodes your encoded data to tree.
    public TreeNode deserialize(String data) {
        String[] tok = data.split(",");
        Queue<String> q = new LinkedList<>(Arrays.asList(tok));
        return helper2(q);
    }

    private TreeNode helper2(Queue<String> q){
        if(q.isEmpty()){
            return null;
        }
        String tok = q.poll();
        if(tok.equals("#")){
            return null;
        }
        TreeNode node = new TreeNode(Integer.parseInt(tok));
        node.left = helper2(q);
        node.right = helper2(q);
        return node;
    }
}

// Your Codec object will be instantiated and called as such:
// Codec ser = new Codec();
// Codec deser = new Codec();
// TreeNode ans = deser.deserialize(ser.serialize(root));

File: ProblemStatements/0449. SerializeAndDeserializeBST/serializeAndDeserialize.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

type Codec struct {
}

func Constructor() Codec {
	return Codec{}
}

// Serializes a tree to a single string.
func (this *Codec) serialize(root *TreeNode) string {
	var res []string
	var dfs func(*TreeNode)
	dfs = func(n *TreeNode) {
		if n == nil {
			return
		}
		res = append(res, strconv.Itoa(n.Val))
		dfs(n.Left)
		dfs(n.Right)
	}
	dfs(root)
	return strings.Join(res, " ")
}

// Deserializes your encoded data to tree.
func (this *Codec) deserialize(data string) *TreeNode {
	if data == "" {
		return nil
	}
	v := strings.Split(data, " ")
	val := make([]int, len(v))
	for i, s := range v {
		val[i], _ = strconv.Atoi(s)
	}
	id := 0
	var build func(int, int) *TreeNode
	build = func(l, h int) *TreeNode {
		if id >= len(val) {
			return nil
		}
		a := val[id]
		if a < l || a > h {
			return nil
		}
		id++
		n := &TreeNode{Val: a}
		n.Left = build(l, a)
		n.Right = build(a, h)
		return n
	}
	return build(-1<<31, 1<<31-1)
}

/**
 * Your Codec object will be instantiated and called as such:
 * ser := Constructor()
 * deser := Constructor()
 * tree := ser.serialize(root)
 * ans := deser.deserialize(tree)
 * return ans
 */